builtin/providers/github/resource_github_membership.go

```diff
@@ -19,26 +22,94 @@ func resourceGithubMembership() *schema.Resource {
 				ForceNew: true,
 			},
 			"role": &schema.Schema{
-				Type:     schema.TypeString,
-				Required: true,
-				//TODO validate function (member, admin)
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validateRoleValue,
 			},
 		},
 	}
 }
 
 func resourceGithubMembershipCreate(d *schema.ResourceData, meta interface{}) error {
-	return nil
+	client := meta.(*GithubClient).client
+	n := d.Get("username").(string)
+	r := d.Get("role").(string)
+
+	membership, _, err := client.Organizations.EditOrgMembership(n, meta.(*GithubClient).organization,
+		&github.Membership{Role: &r})
+	if err != nil {
+		return err
+	}
+
+	orgName := membership.Organization.Login
+	userName := membership.User.Login
+
+	d.SetId(buildMembershipId(*orgName, *userName))
+
+	return resourceGithubMembershipRead(d, meta)
 }
 
 func resourceGithubMembershipRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*GithubClient).client
+
+	membership, _, err := client.Organizations.GetOrgMembership(d.Get("username").(string), meta.(*GithubClient).organization)
+	if err != nil {
+		d.SetId("")
+		return nil
+	}
+	username := membership.User.Login
+	roleName := membership.Role
+
+	d.Set("username", *username)
+	d.Set("role", *roleName)
 	return nil
 }
 
 func resourceGithubMembershipUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*GithubClient).client
+	n := d.Get("username").(string)
+	r := d.Get("role").(string)
+
+	if d.HasChange("role") {
+		_, _, err := client.Organizations.EditOrgMembership(n, meta.(*GithubClient).organization, &github.Membership{
+			Role: &r,
+		})
+		if err != nil {
+			return err
+		}
+	}
 	return nil
 }
 
 func resourceGithubMembershipDelete(d *schema.ResourceData, meta interface{}) error {
-	return nil
+	client := meta.(*GithubClient).client
+	n := d.Get("username").(string)
+
+	_, err := client.Organizations.RemoveOrgMembership(n, meta.(*GithubClient).organization)
+
+	return err
+}
+
+// return the pieces of the id as org, user
+func parseMembershipId(id string) (string, string) {
+	parts := strings.SplitN(id, ":", 2)
+	return parts[0], parts[1]
+}
+
+// Since there is no id for memberships, we are storing in form organization:user
+func buildMembershipId(org, user string) string {
+	return fmt.Sprintf("%s:%s", org, user)
+}
+
+func validateRoleValue(v interface{}, k string) (we []string, errors []error) {
```

Comment:
We should call this `validateMembershipRoleValue` because when we implement the Team memberships, we'll have another copy of this function with different values...

Well, actually, it might be more concise to do some fancy `func` reference when we hit that, so, maybe this isn't an issue.

This could also move to `util.go` as it's more a global function we can reuse.